ngraph/core/reference/include/ngraph/runtime/reference/deformable_convolution.hpp

```diff
@@ -0,0 +1,192 @@
+
+                NGRAPH_CHECK(f_shape.size() == 4, "Unsupported kernel rank: ", f_shape);
+
+                // here we are converting all param types to int's to avoid arithmetic issues
+                // (e.g signed + unsigned) in indexes calculation later
+                ConvolutionParams params{strides, dilation, pads_begin, pads_end};
+
+                // here we are extending spatial dimensions to 3D, because we are going to use 3D
+                // convolution implementation to convolve also in 1D & 2D case
+                Shape input_shape{in_shape};
+                Shape offset_shape(o_shape);
+                Shape filters_shape{f_shape};
+                if (in_shape.size() < 5)
+                {
+                    extend_to_3D(params, input_shape, offset_shape, filters_shape);
+                }
+
+                const size_t batches_count = input_shape[in_batch_axis];
+                const Shape batch_shape(++input_shape.begin(), input_shape.end());
+                const size_t batch_size = shape_size(batch_shape);
+
+                const size_t filters_count = filters_shape[filter_out_ch_axis];
+                const Shape filter_shape(++filters_shape.begin(), filters_shape.end());
+                const size_t filter_size = shape_size(filter_shape);
+
+                auto batch = in;
+                for (size_t batch_idx = 0; batch_idx < batches_count; ++batch_idx)
+                {
+                    auto filter = f;
+                    for (size_t f_idx = 0; f_idx < filters_count; ++f_idx)
+                    {
+                        convolve_3D_channels(params, batch, batch_shape, filter, filter_shape, out);
+                        filter += filter_size;
+                    }
+                    batch += batch_size;
+                }
+            }
+
+            template <typename T>
+            void deformable_convolution(const T* in,
+                                        const T* o,
+                                        const T* f,
+                                        T* out,
+                                        const Shape& in_shape,
+                                        const Shape& o_shape,
+                                        const Shape& f_shape,
+                                        const Shape& out_shape,
+                                        const Strides& strides,
+                                        const Strides& dilation,
+                                        const CoordinateDiff& pads_begin,
+                                        const CoordinateDiff& pads_end,
+                                        const int64_t groups,
+                                        const int64_t deformable_groups)
+
+            {
+                // this implementation supports 2D convolutions
+                NGRAPH_CHECK(in_shape.size() == 4, "Unsupported input rank: ", in_shape);
```

Comment:
Please extract all of these NGRAPH_CHECKs into "validate" helper function. You can borrow one from @ggalieroc  when it is done in #4278 - it should have more extensive checking. The point is to validate all input parameters to ref impl in one place, and later assume they are correct.